### ISSUE TITLE
Make LSystem an overrideable ES6 class

### ISSUE THREAD

**Issue by TMiguelT** [nylki/lindenmayer#13]
I was attempting to use the LSystem class as follows 
```javascript
import LSystem from 'lindenmayer'

export class SimpleTreeSystem extends LSystem {
    //Override the parent constructor to use a specific system
    constructor() {
        super({
            axiom: ['X'],
            productions: {
                X: () => [
                    {symbol: 'F', value: 1},
                    '[',
                    {symbol: '+', value: 20},
                    'X',
                    ']',
                    {symbol: 'F', value: 1},
                    '[',
                    {symbol: '+', value: -20},
                    'X',
                    ']',
                    {symbol: '+', value: 20},
                    'X'
                ],
                F: ({part}) => [{symbol: 'F', value: part.value * 2}]
            }
        })
    }

    /**
     * Override the normal iterate method so that by default it iterates 7 times
     * @param {number} iterations The number of times to iterate the L-System
     */
    iterate(iterations){
        const n = 7
        return super.iterate(iterations || n)
    }
}
```
However since the LSystem function simply sets method properties on the object instead of its prototype, that means that my custom `iterate` implementation is overwritten by the LSystem implementation. 

Could the LSystem function be rewritten to a `class`?

**Comment by nylki** [comment]
Hey again @TMiguelT,
You are keeping me on my toes, I like that. ;)
If your custom iterate function does not need to have the same name, you can use JS' prototype-based inheritance:

```.js
ffunction SimpleTreeSystem(params) {
  LSystem.call(this, params);
  this.superIterate = this.iterate;
  this.iterate = iterations => this.superIterate(iterations || 7);
}

SimpleTreeSystem.prototype = Object.create(LSystem.prototype);
SimpleTreeSystem.prototype.constructor = SimpleTreeSystem;

let test = new SimpleTreeSystem({
  axiom: "F",
	productions: {
		"F": "FF"
	}
})

test.setAxiom("F");
console.log(test.iterate()); // outputs FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF

```
See also: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain

**Comment by nylki** [comment]
I realize, that ES6 class syntax looks much nicer and makes some things more easy and I actually considered using ES6 classes in the beginning. I decided against it because I didn't see the benefit over existing JS object inheritance which I began to really like and wanted to keep things simple.  ES6 classes are after all only syntactical sugar on top of prototype-based inheritance and transpile back into it.

However, this is a topic I am really split about. I really like the better readability of the new class syntax and it would make some things easier, like in your example.
But I am also aware of [arguments made against the use of the class syntax in its current incarnation ](https://github.com/joshburgess/not-awesome-es6-classes) although not all apply in our case of course.

Again, such a long reply, but I feel that it can only help to make my initial motivations and reasonings clear to you and other users of this library.

Having said that if you think it does make sense to use ES6 class syntax here and give me enough arguments to convince me, please fork the repo and create a pull-request for me to review if you like :)

**Comment by TMiguelT** [comment]
Yep! I'm happy to write a PR for this (it might take me a bit to get around to it!). 

I could use the method you've shown, but that would mean not treating my class as a class. And I don't think  `SimpleTreeSystem.prototype = Object.create(LSystem.prototype);` will have any effect, since the LSystem.prototype actually has anything in it.

It seems like you've gone for more of a factory function which I'm not totally sold on. I want to be able to use polymorphism where I write a function that takes an LSystem class or instance, and it calls `iterate` on that instance, meaning that I can override `iterate` in the child class and everything will still work the same.

I'm incredibly unconvinced by the `class` hatred. No doubt classes are less flexible that factories or even manually using prototypes, but they are much easier to write, to read, to understand coming from an OO background, and they make polymorphic behaviour incredibly easy and concise. Yes I could rewrite my code to use a factory but the code would be much longer and much less clear.

**Comment by nylki** [comment]
I am closing this as I don't see the particular need for `class` syntax at this point. Feel free to keep commenting here if you have something to contribute to the discussion.

**Comment by TMiguelT** [comment]
So I take it you don't want `LSystem` to be a class? I'm still happy to write a PR

**Comment by nylki** [comment]
@TMiguelT Hey there! I assumed you lost interest, thats why I closed this. I'll reopen then.

If you if have a clean, working implementation I am happy to evaluate it and possibly merge it, if it meets my expectations.

What I'd additionally like, is to have a proper use case, that make use of the `class`-syntax to warrant the change. This should be something more interesting than overwriting default values. They should be well documented with examples for new users to benefit from. Also, ideally there should be additional unit tests covering the new syntax.

**Comment by TMiguelT** [comment]
I mean, off the top of my head, the advantages of a class would be:

* Users can override specific methods, allowing for better subclassing (very little is possible now). e.g. overriding the string conversion, changing defaults like I've done, adding new methods
* Methods are attached to the prototype rather than to every instance like they are now. This would optimize the memory usage of the library (in situations where the JIT compiler doesn't do that automatically)
* Cleaner syntax. e.g. `this.setAxiom = function (axiom) {}` becomes `setAxiom(axiom){}`
* Integrates better with strong typing systems (e.g. Typescript), which have well defined methods of dealing with classes

**Comment by TMiguelT** [comment]
Closed by #21.
